Codata constants.

MilanSkocic opened the pull request:
The codata constants are implemented as derived types (#347). This way, the values, the uncertainties and the units are easily available. Only the latest release is implemented i.e. 2018. The 2022 CODATA adjustment are not available yet.

The constants, as provided by the NIST, are double precision reals and I though it was useless to provide to include any other  precision.

Your comments and suggestions are welcome.

jvdp1 commented:
Related with:
* #347
* [Discourse](https://fortran-lang.discourse.group/t/physical-constants/822)
* https://gitlab.com/apthorpe/stdlib-codata/-/blob/1c541b3a71556fd89cc6a9f03d401c757d182ceb/src/STDLIB_CODATA_2018.f90

jvdp1 commented:
Thank you @MilanSkocic for this PR.
I am not a user of such constants. So I will add a few possible reviewers.

A general question/comment is about the kinds provided. I think we should try to provide at least single and double precision (it should be easy with `fypp`). But it is maybe useless (I don't know as I never use such things).

jalvesz commented:
Just a style comment. Maybe instead of defining the actual value of the constants with `real(7294.29954142d0,dp)` you could simply use `7294.29954142_dp`which already declares the kind from the parameter `dp`, instead of asking for a conversion it the same kind.

MilanSkocic commented:
> Thank you @MilanSkocic for this PR. I am not a user of such constants. So I will add a few possible reviewers.
> 
> A general question/comment is about the kinds provided. I think we should try to provide at least single and double precision (it should be easy with `fypp`). But it is maybe useless (I don't know as I never use such things).

I have a dedicated for generating the constants for the stdlib (https://github.com/MilanSkocic/stdlib_codata) where I have tested single precision. However, I've got overflows for some constants. I think that for the codata constants the single precision is not relevant.

MilanSkocic commented:
> Just a style comment. Maybe instead of defining the actual value of the constants with `real(7294.29954142d0,dp)` you could simply use `7294.29954142_dp` which already declares the kind from the parameter `dp`, instead of asking for a conversion to the same kind.

Right! I will make the change as soon as possible.

perazz commented:
This is a very interesting proposal @MilanSkocic. I have a comment that does not want to mandate any specifications, but just lay out ideas on how this could be made the most productive. 

 have looked at the state of the art as offered from other well-established packages: [Scipy](https://docs.scipy.org/doc/scipy/reference/constants.html#energy) offers a full range of CODATA constants. 
They're very easy to use because they're basically just numeric constants. So I think it would be most useful if `stdlib` could also provide that facility. I think there are two ways to do it: 

1) Implement all overloaded operators (`==`, `/=`, `>`, `<`, `+`, `-`, `*`, `**`, `/`, etc.) so that these `parameter` derived types can be used exactly as any other `real` parameter. However, because they also contain units and other information (i.e. error), more tools could be available e.g. something like `.approx.` to check whether a value is within the uncertainty bounds of the constant.

2) Also provide a "macro-like" interface to each of them, so that they can also be used as a `real` constant in code. for example: 

```fortran
interface gravitational_constant
   module procedure g_sp
   module procedure g_dp
   ...
end interface

!...
elemental real(dp) function g_dp(mold) 
   real(dp), intent(in), optional :: mold ! dp is the default kind
   g_dp = real(NEWTONIAN_CONSTANT_OF_GRAVITATION%value, kind=dp)
end function g_dp

elemental real(dp) function g_sp(mold)
   real(sp), intent(in) :: mold ! sp is not the default kind -> not optional
   g_sp = real(NEWTONIAN_CONSTANT_OF_GRAVITATION%value, kind=sp)
ed function g_sp
!etc.
```
 
so one could use them like

```
real(dp) :: energy = h * gravitational_constant()
real(sp) :: nrg = h * gravitational_constant(mold=0.0)
```

jalvesz commented:
Or maybe use the generic facility to avoid having to declare an interface for each constant? like this: https://godbolt.org/z/17qP4TWoa

perazz commented:
> the generic facility

It's definitely useful if this class has a few methods (including a nice `print`, etc.). For the purpose of just returning the numeric value (and because there is no `constexpr` in Fortran), I think a nice named interface is more readable IMHO. 
Another option would be to override `real` and do something like `real(NEWTONIAN_CONSTANT_OF_GRAVITATION,kind=dp)` which is also Fortranic, but is too long in my opinion

MilanSkocic commented:
@perazz You are right, in [Scipy](https://docs.scipy.org/doc/scipy/reference/constants.html#constants-database), some of the most common physical constants are available directly as reals. They are aliases to the values of the complete list of Codata constants which are implemented as a dictionary where keys are names and items are tuples containing the value, the uncertainty and the unit. 
I can implement the codata constants as derived types with the generic facility as suggested by @jalvesz in a module called `stdlib_codata` and add aliases to the values of the most common physical constants in a module called `stdlib_constants` as it is done in Scipy. In the latter, mathematical constants, such as `PI` (#99), could be added later.

An example of using the constants:
`program test
! use of the aliases to the values of the most common codata constants and mathematical constants
use stdlib_constants, only: c, PI  
! use of other codata constants
use stdlib_codata, only: ALPHA_PARTICLE_ELECTRON_MASS_RATIO !

real :: b
b = ALPHA_PARTICLE_ELECTRON_MASS_RATIO%eval(b) ! generic facility

end program test`

perazz commented on src/stdlib_constants.fypp:
```diff
@@ -0,0 +1,66 @@
+
+    
+    ! Physical constants
+    real(dp), parameter, public :: c = SPEED_OF_LIGHT_IN_VACUUM%value !! Speed of light in vacuum
```
I like the fact that these constants are only stored in 64-bit precision: most of them are accurate enough nowadays that 32-bit accuracy is not enough. But, at the same time, I've never seen anyone that is more than 15-digit accurate. May be worth checking though.

perazz commented on src/stdlib_constants.fypp:
```diff
@@ -0,0 +1,66 @@
+
+    ! mathematical constants
+    #:if WITH_QP
+    real(qp), parameter, public :: PI_qp    = 3.1415926535897932384626433832795028841971_qp !! PI quadruple precision
```
Should a function-based value be preferred where possible? 
For example, for $\pi$, one could have: 
```fortran
real(wp), parameter, public :: PI = acos(-1.0_wp)
real(wp), parameter, public :: GOLDEN_RATIO = 1 + sqrt(5.0_wp)/2
```

jalvesz replied:
That seems about right!

MilanSkocic replied:
Redefined PI as suggested and initialized from acos(-1).

MilanSkocic commented on src/stdlib_constants.fypp:
```diff
@@ -0,0 +1,66 @@
+    use stdlib_kinds, only: sp, dp, qp
+    use stdlib_codata, only: SPEED_OF_LIGHT_IN_VACUUM, &
+                             VACUUM_ELECTRIC_PERMITTIVITY, &
+                             VACUUM_MAG_PERMEABILITY, &
```
I wanted to keep it homogeneous with the names provided by the [NIST](https://physics.nist.gov/cuu/Constants/Table/allascii.txt). I don'k know why, but magnetic is always shortened in all constants.

jalvesz replied:
Right, feel free to omit this review.

MilanSkocic commented on src/stdlib_codata_type.fypp:
```diff
@@ -0,0 +1,50 @@
+#:include "common.fypp"
+#:set KINDS = ["sp", "dp"]
```
Done.

MilanSkocic commented:
Do I need to create the associated documentation in `/docs/specs`?
I think that an example would be useful in the `example `folder. Do you think that a dedicated subfolder `constants `or `codata` is necessary or the example can be placed in a already existing subfolder?

jvdp1 commented:
> Do I need to create the associated documentation in `/docs/specs`? 

Yes, please. It could be called `stdlib_constants.md`

> I think that an example would be useful in the `example `folder. Do you think that a dedicated subfolder `constants `or `codata` is necessary or the example can be placed in an already existing subfolder?

It should be in a dedicated subfolder, e.g., `constants`.

MilanSkocic commented:
> > Do I need to create the associated documentation in `/docs/specs`?
> 
> Yes, please. It could be called `stdlib_constants.md`
> 
> > I think that an example would be useful in the `example `folder. Do you think that a dedicated subfolder `constants `or `codata` is necessary or the example can be placed in an already existing subfolder?
> 
> It should be in a dedicated subfolder, e.g., `constants`.

MilanSkocic commented:
I have added the documentation and the example.

perazz commented on src/stdlib_codata_type.fypp:
```diff
@@ -0,0 +1,47 @@
+
+#:for p in PROPS
+#:for k in KINDS
+elemental real(${k}$) function eval_${p}$_${k}$(self, mold) result(r)
```
It's great to have a mechanism that natively converts the constant DT to a `real` value. For consistency with `stdlib`, would it be nice to have a `to_real` function that lives both inside the derived type and as a separate interface?

https://github.com/fortran-lang/stdlib/blob/c3d847796638eb72283e2bb370ebf426a2a46b13/example/strings/example_to_string.f90#L19


https://github.com/fortran-lang/stdlib/blob/c3d847796638eb72283e2bb370ebf426a2a46b13/example/bitsets/example_bitsets_to_string.f90#L10

MilanSkocic replied:
Great idea. I'll implement it.

MilanSkocic replied:
The function `to_real` would be nice for the value and the uncertainty. It means that we would have 2 functions:
- `to_real_value`
- `to_real_uncertainty`
Or you meant that the function `to_real` would only apply to the value?

MilanSkocic replied:
I'll use an optional argument for getting the uncertainty and the default will be the value of the constant.

```fortran

#:for k in KINDS
elemental real(${k}$) function to_real_${k}$(self, mold, uncertainty) result(r)
    !! Evaluate the constant value or uncertainty ${p}$ for the kind ${k}$
    class(codata_constant), intent(in) :: self
    real(${k}$), intent(in) :: mold
    logical, intent(in), optional :: uncertainty
    logical :: u
    u = optval(uncertainty, .false.)
    if(u .eq. .true.)then
        r = real(self%uncertainty, kind(mold))
    else
        r = real(self%value, kind(mold))
    endif
end function
#:endfor
```
Identical mechanism for the function `to_real` that lives outside of the DT.

perazz replied:
LGTM

MilanSkocic commented:
@perazz `to_real` for consistency with stdlib has been added.

perazz reviewed (approved):
LGTM as far as my suggestions, thank you @MilanSkocic

jvdp1 commented on doc/specs/stdlib_constants.md:
```diff
@@ -0,0 +1,21 @@
+
+The [[stdlib_constants]] module provides mathematical constants and the most common physical constants.
+
+The [[stdlib_codata]] module defines all codata constants as derived type [[codata_constant]].
```
DT have usually the suffix `_type` in `stdlib`:
```suggestion
The [[stdlib_codata]] module defines all codata constants as derived type [[codata_constant_type]].
```

jvdp1 commented on doc/specs/stdlib_constants.md:
```diff
@@ -0,0 +1,21 @@
+The [[stdlib_constants]] module provides mathematical constants and the most common physical constants.
+
+The [[stdlib_codata]] module defines all codata constants as derived type [[codata_constant]].
+The lastest codata constants were released in 2018 by the [NIST](http://physics.nist.gov/constants.)
```
```suggestion
The latest codata constants were released in 2018 by the [NIST](http://physics.nist.gov/constants.)
```

jvdp1 commented on example/constants/example_constants.f90:
```diff
@@ -0,0 +1,25 @@
+program example_constants
+    use stdlib_constants, only: c, PI=>PI_dp
```
Following the [style guide](https://github.com/fortran-lang/stdlib/blob/33559f551d1c7f22ea2511e4f7e779e2aed410c3/STYLE_GUIDE.md?plain=1#L39), variable names should be lowercase.
```suggestion
    use stdlib_constants, only: c, pi=>PI_dp
```
This is also consisten with the following line (`alpha =>...`)

jvdp1 commented on example/constants/example_constants.f90:
```diff
@@ -0,0 +1,25 @@
+    print *, "speed of light in vacuum= ", c
+
+    ! Use of mathematical constants such as PI
+    print *, "PI as double precision real= ", PI
```
```suggestion
    print *, "PI as double precision real= ", pi
```

jvdp1 commented on src/stdlib_codata.f90:
```diff
@@ -0,0 +1,1779 @@
+module stdlib_codata
+    !! Codata Constants - Autogenerated
```
is the code to autogenerate this module available? Please provide a short explanation/reference.

MilanSkocic replied:
Yes the code is available. If needed, it can be incorporated in stdlib.

jvdp1 replied:
OK. A short explanation and a link to the Python script should be sufficient. It is mainly to inform future developers.

jvdp1 commented on src/stdlib_codata_type.fypp:
```diff
@@ -0,0 +1,61 @@
+    use stdlib_optval, only: optval 
+    private
+
+    type, public :: codata_constant
```
```suggestion
    type, public :: codata_constant_type
```
To be in agreement with other DT provided by `stdlib`.

jvdp1 commented on doc/specs/stdlib_constants.md:
```diff
@@ -0,0 +1,21 @@
+The [[stdlib_codata]] module defines all codata constants as derived type [[codata_constant]].
+The lastest codata constants were released in 2018 by the [NIST](http://physics.nist.gov/constants.)
+The names are quite long and can be aliased with shorter names.
+
```
The DT should be described in details here. There should be also a note that all provided values are `dp`.

jvdp1 commented on src/stdlib_codata_type.fypp:
```diff
@@ -0,0 +1,61 @@
+        #:endfor
+    end interface
+
+    public :: to_real
```
```suggestion
    public :: real
```
If I am correct, it could be called `real` and follow the same interface as the intrinsic function [`real`](https://gcc.gnu.org/onlinedocs/gfortran/REAL.html)

MilanSkocic replied:
It could be called `real` but I was wondering if it is a good idea to mask an intrinsic function. Moreover, the name suggested by @perazz made sense to me when I looked at the `stdlib_strings` because the name carries the notion of conversion.

jvdp1 replied:
The intrinsic `real` should not be masked, but just overloaded.
It is also fine with me to use `to_real` (similarly to `to_string`).
@perazz @jalvesz What is your opinion here? do you prefer `real` or `to_real`?

perazz replied:
Both are fine to me - I also had originally thought about `real`, although the overloaded interface would then be different than the intrinsic: 
- `real(x, KIND)` has `integer, intent(in) :: kind` and the fortran compiler resolves that to the return value
- `to_real(x, MOLD)` has `real(*), intent(in) :: mold` because the return kind cannot otherwise be selected.

jvdp1 commented on src/stdlib_codata_type.fypp:
```diff
@@ -0,0 +1,61 @@
+end subroutine
+
+#:for k in KINDS
+elemental pure real(${k}$) function to_real_${k}$(self, mold, uncertainty) result(r)
```
```suggestion
elemental pure real(${k}$) function real_${k}$(self, mold, uncertainty) result(r)
```

jvdp1 replied:
This function should be also described  in the specs

jvdp1 commented on src/stdlib_codata_type.fypp:
```diff
@@ -0,0 +1,61 @@
+
+#:for k in KINDS
+elemental pure real(${k}$) function to_real_${k}$(self, mold, uncertainty) result(r)
+    !! Get the constant value or uncertainty for the kind ${k}$
```
```suggestion
    !! version: experimental
    !!
    !! Get the constant value or uncertainty for the kind ${k}$
    !! [Specification](TO_BE_COMPLETED)
```

jvdp1 commented on src/stdlib_codata_type.fypp:
```diff
@@ -0,0 +1,61 @@
+    private
+
+    type, public :: codata_constant
+    !! Derived type for representing a Codata constant.
```
```suggestion
        !! version: experimental
        !!
        !! Derived type for representing a Codata constant.
        !! [Specification](TOBECOMPLETED)
```

jvdp1 reviewed (commented):
Thank you @MilanSkocic for this PR. I think it is close to be merged. Here are some minor comments.
Could you maybe also add some tests? At least for the DR?

MilanSkocic commented:
> Thank you @MilanSkocic for this PR. I think it is close to be merged. Here are some minor comments. Could you maybe also add some tests? At least for the DR?

No problem the tests. I was already working on. 

The code for autogenerating the `stdlib_codata` module is available [here](https://github.com/MilanSkocic/stdlib_codata). It is a simple parser written in Python of the raw data provided by the NIST. If needed it can be incorporated in stdlib.

MilanSkocic commented:
I added tests for some Codata constants (values and uncertainties) that I have arbitrary chosen. 

I did not implement any test for mathematical constants as I have only defined one (PI).  I am not sure how the mathematical constants should be  implemented i.e. either the way I did or through derived type as mentioned issue #99.
@jvdp1 Let me know if you want me to remove the mathematical constant `PI` until a definitive decision is made.

jvdp1 commented on doc/specs/stdlib_constants.md:
```diff
@@ -0,0 +1,72 @@
+
+* 2 type-bound procedures:
+
+    * `print` for print the values of the constant members.
```
```suggestion
    * `print`: to print the values of the constant members;
```

jvdp1 commented on doc/specs/stdlib_constants.md:
```diff
@@ -0,0 +1,72 @@
+* 2 type-bound procedures:
+
+    * `print` for print the values of the constant members.
+    * `to_real` for getting the value or the uncertainty to the desired precision.
```
```suggestion
    * `to_real`: to get the value or the uncertainty to the desired precision.
```

jvdp1 commented on doc/specs/stdlib_constants.md:
```diff
@@ -0,0 +1,72 @@
+A module level interface [[stdlib_codata_type(module):to_real(interface)]] is 
+available for getting the constant value or uncertainty of a constant. 
+
+## `to_real` - Get the constant value or uncertainty.
```
```suggestion
## `to_real` - Get the constant value or its uncertainty.
```

jvdp1 commented on doc/specs/stdlib_constants.md:
```diff
@@ -0,0 +1,72 @@
+
+### Description
+
+Convert a [[stdlib_codata_type(module):codata_constant_type(type)]] to a real. 
```
```suggestion
Convert a [[stdlib_codata_type(module):codata_constant_type(type)]] to a `real` (at least `sp`, or `dp`) scalar. 
```

jvdp1 commented on doc/specs/stdlib_constants.md:
```diff
@@ -0,0 +1,72 @@
+### Description
+
+Convert a [[stdlib_codata_type(module):codata_constant_type(type)]] to a real. 
+**Warning**: Some constants cannot be converted to simple precision reals due to the values of the exponents.
```
```suggestion
**Warning**: Some constants cannot be converted to single precision `sp` reals due to the value of the exponents.
```

jvdp1 commented on test/constants/test_constants.f90:
```diff
@@ -0,0 +1,408 @@
+                             PLANCK_CONSTANT,&
+                             SPEED_OF_LIGHT_IN_VACUUM,&
+                             STANDARD_ACCELERATION_OF_GRAVITY
+
```
```suggestion
    implicit none
```

jvdp1 commented on test/constants/test_constants.f90:
```diff
@@ -0,0 +1,408 @@
+                             SPEED_OF_LIGHT_IN_VACUUM,&
+                             STANDARD_ACCELERATION_OF_GRAVITY
+
+    public :: collect_constants
```
```suggestion
    private
    public :: collect_constants
```

jvdp1 commented on test/constants/test_constants.f90:
```diff
@@ -0,0 +1,408 @@
+
+!> Collect all exported unit tests
+subroutine collect_constants(testsuite)
+    implicit none
```
```suggestion
```

No need to mention `implicit none` here if it is mentioned for the whole module

jvdp1 commented on test/constants/test_constants.f90:
```diff
@@ -0,0 +1,408 @@
+end subroutine
+
+subroutine test_year(error)
+    implicit none
```
```suggestion
```

jvdp1 commented on test/constants/test_constants.f90:
```diff
@@ -0,0 +1,408 @@
+end subroutine
+
+subroutine test_ALPHA_PARTICLE_ELECTRON_MASS_RATIO(error)
+    implicit none
```
```suggestion
```

jvdp1 commented on test/constants/test_constants.f90:
```diff
@@ -0,0 +1,408 @@
+end subroutine
+
+subroutine test_ALPHA_PARTICLE_MASS(error)
+    implicit none
```
```suggestion
```

jvdp1 commented on test/constants/test_constants.f90:
```diff
@@ -0,0 +1,408 @@
+end subroutine
+
+subroutine test_ATOMIC_MASS_CONSTANT(error)
+    implicit none
```
```suggestion
```

jvdp1 commented on test/constants/test_constants.f90:
```diff
@@ -0,0 +1,408 @@
+
+subroutine test_ALPHA_PARTICLE_ELECTRON_MASS_RATIO(error)
+    implicit none
+    type(error_type), allocatable, intent(out) :: error 
+    real(dp) :: value, expected, diff, fac
+    fac = 1.0d0
+    expected = 7294.29954171d0 * fac
+    value = ALPHA_PARTICLE_ELECTRON_MASS_RATIO%to_real(fac) * fac
+    diff = expected - value
+    call check(error, diff, 0.0d0)
```
```suggestion
    type(error_type), allocatable, intent(out) :: error 
    real(dp), parameter :: fac = 1.d0
    
    call check(error, 7294.29954171d0 * fac - ALPHA_PARTICLE_ELECTRON_MASS_RATIO%to_real(fac) * fac, 0.0d0)
```
Just a suggestion

MilanSkocic replied:
Good idea. I'll make it more compact.

jvdp1 commented on test/constants/test_constants.f90:
```diff
@@ -0,0 +1,408 @@
+end subroutine
+
+subroutine test_AVOGADRO_CONSTANT(error)
+    implicit none
```
```suggestion
```

jvdp1 commented on test/constants/test_constants.f90:
```diff
@@ -0,0 +1,408 @@
+end subroutine
+
+subroutine test_BOLTZMANN_CONSTANT(error)
+    implicit none
```
```suggestion
```

jvdp1 commented on test/constants/test_constants.f90:
```diff
@@ -0,0 +1,408 @@
+end subroutine
+
+subroutine test_ELECTRON_VOLT(error)
+    implicit none
```
```suggestion
```

jvdp1 commented on test/constants/test_constants.f90:
```diff
@@ -0,0 +1,408 @@
+end subroutine
+
+subroutine test_ELEMENTARY_CHARGE(error)
+    implicit none
```
```suggestion
```

jvdp1 commented on test/constants/test_constants.f90:
```diff
@@ -0,0 +1,408 @@
+end subroutine
+
+subroutine test_FARADAY_CONSTANT(error)
+    implicit none
```
```suggestion
```

jvdp1 commented on test/constants/test_constants.f90:
```diff
@@ -0,0 +1,408 @@
+end subroutine
+
+subroutine test_MOLAR_MASS_CONSTANT(error)
+    implicit none
```
```suggestion
```

jvdp1 commented on test/constants/test_constants.f90:
```diff
@@ -0,0 +1,408 @@
+end subroutine
+
+subroutine test_MOLAR_VOLUME_NTP(error)
+    implicit none
```
```suggestion
```

jvdp1 commented on test/constants/test_constants.f90:
```diff
@@ -0,0 +1,408 @@
+end subroutine
+
+subroutine test_PLANCK_CONSTANT(error)
+    implicit none
```
```suggestion
```

jvdp1 commented on test/constants/test_constants.f90:
```diff
@@ -0,0 +1,408 @@
+end subroutine
+
+subroutine test_SPEED_OF_LIGHT(error)
+    implicit none
```
```suggestion
```

jvdp1 commented on test/constants/test_constants.f90:
```diff
@@ -0,0 +1,408 @@
+end subroutine
+
+subroutine test_STANDARD_ACCELERATION_OF_GRAVITY(error)
+    implicit none
```
```suggestion
```

jvdp1 commented on test/constants/test_constants.f90:
```diff
@@ -0,0 +1,408 @@
+end subroutine
+
+subroutine test_U_ALPHA_PARTICLE_ELECTRON_MASS_RATIO(error)
+    implicit none
```
```suggestion
```

jvdp1 commented on test/constants/test_constants.f90:
```diff
@@ -0,0 +1,408 @@
+end subroutine
+
+subroutine test_U_ALPHA_PARTICLE_MASS(error)
+    implicit none
```
```suggestion
```

jvdp1 commented on test/constants/test_constants.f90:
```diff
@@ -0,0 +1,408 @@
+end subroutine
+
+subroutine test_U_ATOMIC_MASS_CONSTANT(error)
+    implicit none
```
```suggestion
```

jvdp1 commented on test/constants/test_constants.f90:
```diff
@@ -0,0 +1,408 @@
+end subroutine
+
+subroutine test_U_AVOGADRO_CONSTANT(error)
+    implicit none
```
```suggestion
```

jvdp1 commented on test/constants/test_constants.f90:
```diff
@@ -0,0 +1,408 @@
+end subroutine
+
+subroutine test_U_BOLTZMANN_CONSTANT(error)
+    implicit none
```
```suggestion
```

jvdp1 commented on test/constants/test_constants.f90:
```diff
@@ -0,0 +1,408 @@
+end subroutine
+
+subroutine test_U_ELECTRON_VOLT(error)
+    implicit none
```
```suggestion
```

jvdp1 commented on test/constants/test_constants.f90:
```diff
@@ -0,0 +1,408 @@
+end subroutine
+
+subroutine test_U_ELEMENTARY_CHARGE(error)
+    implicit none
```
```suggestion
```

jvdp1 commented on test/constants/test_constants.f90:
```diff
@@ -0,0 +1,408 @@
+end subroutine
+
+subroutine test_U_FARADAY_CONSTANT(error)
+    implicit none
```
```suggestion
```

jvdp1 commented on test/constants/test_constants.f90:
```diff
@@ -0,0 +1,408 @@
+end subroutine
+
+subroutine test_U_MOLAR_MASS_CONSTANT(error)
+    implicit none
```
```suggestion
```

jvdp1 commented on test/constants/test_constants.f90:
```diff
@@ -0,0 +1,408 @@
+end subroutine
+
+subroutine test_U_MOLAR_VOLUME_NTP(error)
+    implicit none
```
```suggestion
```

jvdp1 commented on test/constants/test_constants.f90:
```diff
@@ -0,0 +1,408 @@
+end subroutine
+
+subroutine test_U_PLANCK_CONSTANT(error)
+    implicit none
```
```suggestion
```

jvdp1 commented on test/constants/test_constants.f90:
```diff
@@ -0,0 +1,408 @@
+end subroutine
+
+subroutine test_U_SPEED_OF_LIGHT(error)
+    implicit none
```
```suggestion
```

jvdp1 commented on test/constants/test_constants.f90:
```diff
@@ -0,0 +1,408 @@
+end subroutine
+
+subroutine test_U_STANDARD_ACCELERATION_OF_GRAVITY(error)
+    implicit none
```
```suggestion
```

jvdp1 commented on test/constants/test_constants.f90:
```diff
@@ -0,0 +1,408 @@
+    implicit none
+    type(error_type), allocatable, intent(out) :: error 
+    real(dp) :: value, expected, diff, fac
+    fac = 1.0d0
```
Suggestion: `fac` could be a `parameter` at the level of the `module`. So it should declared only once.

MilanSkocic replied:
I changed `fac` as a module level variable. I  did not declare it as a parameter because I need to able to change according to the values that are being tested.

jvdp1 reviewed (commented):
Thank you @milanskocic for the recent commits. I think it is almost ready to be merged. I only have some minor edits.
Once considered, I will wait a couple of days for potential additional reviews and then merge it.

jvdp1 commented:
Could you also rebase your code to solve the conflicts, please?

MilanSkocic commented:
No problem @jvdp1 . I will take into account all your comments and suggestions. 
I will solve the conflicts and I will let you know when it is done.

jvdp1 commented:
Sorry. I though it was always the same value. If not, it should not be a
module variable because the tests can be run in parallel.
Could it be a mix of the two approaches?

Le dim. 19 mai 2024, 07:38, Milan Skocic ***@***.***> a
écrit :

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In test/constants/test_constants.f90
> <https://github.com/fortran-lang/stdlib/pull/800#discussion_r1605951192>:
>
> > +    implicit none
> +    type(error_type), allocatable, intent(out) :: error
> +    real(dp) :: value, expected, diff, fac
> +    fac = 1.0d0
> +    expected = 0.0d0 * fac
> +    value = SPEED_OF_LIGHT_IN_VACUUM%to_real(fac, uncertainty=.true.) * fac
> +    diff = expected - value
> +    call check(error, diff, 0.0d0)
> +    if (allocated(error)) return
> +end subroutine
> +
> +subroutine test_U_STANDARD_ACCELERATION_OF_GRAVITY(error)
> +    implicit none
> +    type(error_type), allocatable, intent(out) :: error
> +    real(dp) :: value, expected, diff, fac
> +    fac = 1.0d0
>
> I changed fac as a module level variable. I did not declare it as a
> parameter because I need to able to change according to the values that are
> being tested.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/fortran-lang/stdlib/pull/800#discussion_r1605951192>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD5RO7EE4AY2HZ3N6CO767TZDA3HHAVCNFSM6AAAAABGRC7NM6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDANRUHE4TOMZSGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

MilanSkocic commented:
> Sorry. I though it was always the same value. If not, it should not be a
> 
> module variable because the tests can be run in parallel.
> 
> Could it be a mix of the two approaches?
> 
> 
> 
> Le dim. 19 mai 2024, 07:38, Milan Skocic ***@***.***> a
> 
> écrit :
> 
> 
> 
> > ***@***.**** commented on this pull request.
> 
> > ------------------------------
> 
> >
> 
> > In test/constants/test_constants.f90
> 
> > <https://github.com/fortran-lang/stdlib/pull/800#discussion_r1605951192>:
> 
> >
> 
> > > +    implicit none
> 
> > +    type(error_type), allocatable, intent(out) :: error
> 
> > +    real(dp) :: value, expected, diff, fac
> 
> > +    fac = 1.0d0
> 
> > +    expected = 0.0d0 * fac
> 
> > +    value = SPEED_OF_LIGHT_IN_VACUUM%to_real(fac, uncertainty=.true.) * fac
> 
> > +    diff = expected - value
> 
> > +    call check(error, diff, 0.0d0)
> 
> > +    if (allocated(error)) return
> 
> > +end subroutine
> 
> > +
> 
> > +subroutine test_U_STANDARD_ACCELERATION_OF_GRAVITY(error)
> 
> > +    implicit none
> 
> > +    type(error_type), allocatable, intent(out) :: error
> 
> > +    real(dp) :: value, expected, diff, fac
> 
> > +    fac = 1.0d0
> 
> >
> 
> > I changed fac as a module level variable. I did not declare it as a
> 
> > parameter because I need to able to change according to the values that are
> 
> > being tested.
> 
> >
> 
> > —
> 
> > Reply to this email directly, view it on GitHub
> 
> > <https://github.com/fortran-lang/stdlib/pull/800#discussion_r1605951192>,
> 
> > or unsubscribe
> 
> > <https://github.com/notifications/unsubscribe-auth/AD5RO7EE4AY2HZ3N6CO767TZDA3HHAVCNFSM6AAAAABGRC7NM6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDANRUHE4TOMZSGA>
> 
> > .
> 
> > You are receiving this because you were mentioned.Message ID:
> 
> > ***@***.***>
> 
> >
> 
> 

No problem.

awvwgk commented on test/constants/test_constants.f90:
```diff
@@ -0,0 +1,382 @@
+    expected = 2022
+    value = YEAR
+    diff = expected - value
+    call check(error, diff, 0)
```
I would recommend to do the comparison between the values directly here instead of using the precomputed difference. The error message from the test framework can be more meaningful this way.

MilanSkocic replied:
No problem. I'll  do the comparison directly as also suggested by @jvdp1 .

awvwgk commented on src/stdlib_constants.fypp:
```diff
@@ -0,0 +1,64 @@
+    #:endfor
+    
+    ! Physical constants
+    real(dp), parameter, public :: c = SPEED_OF_LIGHT_IN_VACUUM%value !! Speed of light in vacuum
```
I would be careful with one letter variables or constants. There are many physical constants and we would not be able to provide their short forms in Fortran nicely because the same letter might be used for several constants. Also, because of the case insensitivity of the language these names might bring surprises for users.

MilanSkocic replied:
The idea was to mimic the constants module as it is done in Scipy. Nonetheless, I'm aware of the potential problem with one letter constants. If everybody agrees I can remove them.

jvdp1 replied:
Instead of removing them, could we think to add a prefix, e.g.,?
```fortran 
... stdlib_pi = ...
... stdlib_c = ...
```

perazz replied:
When using numeric constants, the shorter their names the better imho. Because they're inside a module that only contains them, I think they're mostly harmless. In case one has overlapping names, they can always rename them as: 

```fortran
use stdlib_constants, only: clight => c
```

Of course, it's important that their names are not confusing as @awvwgk is suggesting. But here they're taken from Scipy, so, probably a good reference for most people coming from the Python world.

jvdp1 replied:
I am in favour of keeping the same notation as in Scipy.
@MilanSkocic Could you maybe add a warning in the specs, to highlight this potential issue and how to solve it (@perazz 's suggestion), please?

MilanSkocic replied:
Ok, no problem.

MilanSkocic replied:
Done.

MilanSkocic commented:
@jvdp1 I have implemented the direct comparison in the tests. I rebased my code and solved the conflicts. On my fork, I see only  my commits but in the PR additional commits are present. I might have messed up something when I rebase the code.

jvdp1 commented:
Faire point @perazz

I will not be a user of such data. But i think that they should be kept in
stdlib. Both notations are fine for me

Le mar. 21 mai 2024, 17:43, Federico Perini ***@***.***> a
écrit :

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/stdlib_constants.fypp
> <https://github.com/fortran-lang/stdlib/pull/800#discussion_r1608552969>:
>
> > +                             STEFAN_BOLTZMANN_CONSTANT, &
> +                             WIEN_WAVELENGTH_DISPLACEMENT_LAW_CONSTANT, &
> +                             RYDBERG_CONSTANT, &
> +                             ELECTRON_MASS, &
> +                             PROTON_MASS, &
> +                             NEUTRON_MASS, &
> +                             ATOMIC_MASS_CONSTANT
> +    private
> +
> +    ! mathematical constants
> +    #:for k in KINDS
> +    real(${k}$), parameter, public :: PI_${k}$ = acos(-1.0_${k}$) !! PI
> +    #:endfor
> +
> +    ! Physical constants
> +    real(dp), parameter, public :: c = SPEED_OF_LIGHT_IN_VACUUM%value !! Speed of light in vacuum
>
> When using numeric constants, the shorter their names the better imho.
> Because they're inside a module that only contains them, I think they're
> mostly harmless. In case one has overlapping names, they can always rename
> them as:
>
> use stdlib_constants, only: clight => c
>
> Of course, it's important that their names are not confusing as @awvwgk
> <https://github.com/awvwgk> is suggesting. But here they're taken from
> Scipy, so, probably a good reference for most people coming from the Python
> world.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/fortran-lang/stdlib/pull/800#discussion_r1608552969>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD5RO7DZ36D3LZGQH6ER4JTZDNTSLAVCNFSM6AAAAABGRC7NM6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDANRYHE4TIMRTGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

MilanSkocic commented:
Do we have an agreement to proceed without any prefix on the constants as they are defined in Scipy? @jvdp1

jvdp1 commented:
Thank you @MilanSkocic 
What do you think @perazz @jalvesz @awvwgk ?

jalvesz commented:
LGTM!! brilliant work @MilanSkocic

jvdp1 commented:
thank you @MilanSkocic  for this PR. I will merge it.

@MilanSkocic Don't hesitate to advertise this new addition on Discourse and other channels.

MilanSkocic commented:
Thank you @jvdp1 ! I'm glad the codata constants have been merged. For sure I'll advertise the addition.
It was a pleasure to contribute to the Fortran-lang community.
